server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala

```diff
@@ -0,0 +1,240 @@
+/*
+ * Licensed to Cloudera, Inc. under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  Cloudera, Inc. licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.cloudera.livy.utils
+
+import java.util.concurrent.{ThreadFactory, TimeoutException}
+import java.util.concurrent.Executors
+
+import scala.annotation.tailrec
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.{blocking, Await, ExecutionContext, Future}
+import scala.concurrent.duration._
+import scala.language.postfixOps
+import scala.util.{Failure, Success, Try}
+
+import org.apache.hadoop.yarn.api.records.{ApplicationId, ApplicationReport, FinalApplicationStatus, YarnApplicationState}
+import org.apache.hadoop.yarn.client.api.YarnClient
+import org.apache.hadoop.yarn.conf.YarnConfiguration
+
+import com.cloudera.livy.Logging
+import com.cloudera.livy.util.LineBufferedProcess
+
+object SparkYarnApp extends Logging {
+  private class NonDaemonThreadFactory extends ThreadFactory {
+    def newThread(r: Runnable): Thread = {
+      val t = new Thread(r)
+      t.setDaemon(true)
+      t
+    }
+  }
+
+  // Create a non daemon thread pool to run getAppIdFromTagAsync().
+  // getAppIdFromTagAsync() might take a while and should not use the global default thread pool.
+  private implicit val ec = ExecutionContext.fromExecutor(
+    Executors.newCachedThreadPool(new NonDaemonThreadFactory()))
+
+  // It takes at least 5 seconds to see the newly submitted app.
+  private val APP_TAG_TO_ID_TIMEOUT = 30 seconds
+  private val KILL_TIMEOUT = APP_TAG_TO_ID_TIMEOUT
+  private val POLL_INTERVAL = 1 second
+
+  // YarnClient is thread safe. Create once, share it across threads.
+  private[this] lazy val yarnClient = YarnClient.createYarnClient()
+  private[this] var yarnClientCreated = false
+
+  def getYarnClient(): YarnClient = synchronized {
+    if (!yarnClientCreated) {
+      yarnClient.init(new YarnConfiguration())
+      yarnClient.start()
+      yarnClientCreated = true
+    }
+    yarnClient
+  }
+
+  def getAppIdFromTagAsync(appTag: String): Future[ApplicationId] =
```

Comment:
nit: `Async` is implicit by the method returning a `Future`, so no need to be in the name.